### ISSUE TITLE
How to Attended Transfer?

### ISSUE THREAD

**Issue by daichan4649** [sipsorcery/sipsorcery#121]
I am very grateful for listening to my wishes.

Are there any plans to implement "Attended Transfer" in the softphone sample? Please let me know if you have a plan. It seems to be possible with reference to hold and blind transfer, but it is complicated and difficult.

[Attended Transfer (RFC5589)](https://tools.ietf.org/html/rfc5589)

Also, can I write here if I have the required features or samples? I would like to write there if there is a forum.

**Comment by sipsorcery** [comment]
Yes attended transfer isn't that much work. The building blocks are there it just needs to be wired up. The steps are:

 - Establish call,
 - Attended transfer steps:
  - Place remote party on hold,
  - Place new call to the transfer destination,
  - Bridge first and seconds calls together with a REFER request.

I'll work on it once the current softphone enhancements are done.

> Also, can I write here if I have the required features or samples? I would like to write there if there is a forum.

If it's a full program along the lines of the existing [examples](https://github.com/sipsorcery/sipsorcery/tree/master/examples) you could submit a Pull Request.

If it's more of an explanatory or teaching article, like [these](https://sipsorcery.github.io/sipsorcery/articles/intro.html), then you can write it in markdown and submit a Pull Request on the `gh-pages` branch. Markdown for the articles can be found [here](https://github.com/sipsorcery/sipsorcery/tree/gh-pages/docfx/articles) (note it needs to be markdown not html, the html files are auto-generated from the markdown).

**Comment by daichan4649** [comment]
Sorry for the late reply.

I see. Thank you for the steps example. thank you for the advice. I'll try it. 

And thank you very much for explaining the PR procedure!

**Comment by daichan4649** [comment]
I have a question. For example, in the following scenario:

- UA1 and UA2 are currently talking
- UA1 transfers UA2 to UA3

Will the logic look roughly like this? Is this the implementation you expect?

### Place remote party on hold
```
var ua = new SIPUserAgent(m_sipTransport, null);

(Incoming, Accept, ...)
(Talking UA1 <-> UA2)

// Hold (UA1 <-> UA2)
ua.PutOnHold();
```
### Place new call to the transfer destination
```
// Call (UA1 --> UA3)
var uaForTransfer = new SIPUserAgent(m_sipTransport, null);
uaForTransfer.Call('UA3');
```
### Bridge first and seconds calls together with a re-INVITE request.
```
// Hold (UA1 <-> UA3)
uaForTransfer.PutOnHold();

// Transfer (UA2 --> UA3)
ua.Transfer('UA3');

// (Finally)
// UA1: Idle
// UA2 <-> UA3: Talking
```

**Comment by sipsorcery** [comment]
Yes that's very close to how it will work. In the last step there's no need to put UA3 on hold. 

I'm also yet to think about the way the `SIPUserAgent` class will bridge a call between two instances or whether it will need two instances. Maybe one instance can handle two dialogs. The optimal choice will become apparent once I start coding which should be very soon now.

**Comment by daichan4649** [comment]
> Yes that's very close to how it will work. In the last step there's no need to put UA3 on hold.
Thank you for the advice.

> The optimal choice will become apparent once I start coding which should be very soon now.
Really?! I expect very much. I am looking forward to it!!

**Comment by sipsorcery** [comment]
The softphone now has an example of how to do attended transfers (I ended up using a `REFER` request rather than a `re-INVITE`).

To see the function you need to make or receive two calls on the softphone. Once the second call is established an `Attended Transfer` button will appear.

**Comment by daichan4649** [comment]
I read the Softphone sample and SIPUSerAgent. Thank you for making it very easy to use! I will use it immediately.

There is one question.

> The softphone now has an example of how to do attended transfers (I ended up using a REFER request rather than a re-INVITE).

A pattern that uses REFER directly (not REFER + re-INVITE) was not found in the RFC. Is this defined somewhere? There is no problem in operation, but I am interested.

**Comment by sipsorcery** [comment]
> I couldn't find a pattern (not REFER + re-INVITE) that uses REFER directly in the RFC. Is this defined somewhere? There is no problem in operation, but I am interested.

When some SIP servers (Asterisk, FreeSWITCH etc) do attended transfers no `REFER` request is used. Instead they will use the re-INVITE mechanism that changes which RTP endpoint the call participants are using. The servers use this mechanism so that they can stay in the SIP signalling path. When the transfer participants hang up the `BYE` requests get sent to the server and allow the generation of Call Detail Records.

The last time I implemented attended transfers was for a server and I looked over that code and thought I'd do i the same way for the softphone. But the softphone does not want to stay in the SIP signalling path it made no sense to use the same re-INVITE mechanism.
